ompi/mca/op/cuda/op_cuda_functions.c

```diff
@@ -0,0 +1,1897 @@
+        source1_rc = opal_accelerator.check_addr(*source1, source1_device, &source1_flags);
+        *device = *target_device;
+
+        if (NULL != orig_source2) {
+            source2_rc = opal_accelerator.check_addr(*source2, source2_device, &source2_flags);
+        }
+
+        if (0 == target_rc && 0 == source1_rc && 0 == source2_rc) {
+            /* no buffers are on any device, select device 0 */
+            *device = 0;
+        } else if (*target_device == -1) {
+            if (*source1_device == -1 && NULL != orig_source2) {
+                *device = *source2_device;
+            } else {
+                *device = *source1_device;
+            }
+        }
+
+        if (0 == target_rc || 0 == source1_rc || *target_device != *source1_device) {
+            size_t nbytes;
+            ompi_datatype_type_size(dtype, &nbytes);
+            nbytes *= count;
+
+            if (0 == target_rc) {
+                // allocate memory on the device for the target buffer
+                opal_accelerator.mem_alloc_stream(*device, target, nbytes, stream);
+                CHECK(cuMemcpyHtoDAsync, ((CUdeviceptr)*target, orig_target, nbytes, *(CUstream*)stream->stream));
+                *target_device = -1; // mark target device as host
+            }
+
+            if (0 == source1_rc || *device != *source1_device) {
+                // allocate memory on the device for the source buffer
+                opal_accelerator.mem_alloc_stream(*device, source1, nbytes, stream);
+                if (0 == source1_rc) {
+                    /* copy from host to device */
+                    CHECK(cuMemcpyHtoDAsync, ((CUdeviceptr)*source1, orig_source1, nbytes, *(CUstream*)stream->stream));
+                } else {
+                    /* copy from one device to another device */
+                    /* TODO: does this actually work? Can we enable P2P? */
+                    CHECK(cuMemcpyDtoDAsync, ((CUdeviceptr)*source1, (CUdeviceptr)orig_source1, nbytes, *(CUstream*)stream->stream));
+                }
+            }
+
+        }
+        if (NULL != source2_device && *target_device != *source2_device) {
+            // allocate memory on the device for the source buffer
+            size_t nbytes;
+            ompi_datatype_type_size(dtype, &nbytes);
+            nbytes *= count;
+
+            opal_accelerator.mem_alloc_stream(*device, source2, nbytes, stream);
+            if (0 == source2_rc) {
+                /* copy from host to device */
+                //printf("copying source from host to device %d\n", *device);
+                CHECK(cuMemcpyHtoDAsync, ((CUdeviceptr)*source2, orig_source2, nbytes, *(CUstream*)stream->stream));
+            } else {
+                /* copy from one device to another device */
+                /* TODO: does this actually work? Can we enable P2P? */
+                CHECK(cuMemcpyDtoDAsync, ((CUdeviceptr)*source2, (CUdeviceptr)orig_source2, nbytes, *(CUstream*)stream->stream));
```

Comment:
I guess this should be guarded with a check for `cudaDeviceCanAccessPeer`?

Reply:
The way I understand the dox in https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__PEER.html is that I need to check for `cudaDeviceCanAccessPeer` and then explicitly enable peer access via `cudaDeviceEnablePeerAccess`. Is that something we can up front during init? Are there any down-sides to enabling peer access?

Reply:
What about this? The docs are unclear and I need some input here. Is it safe to access peer memory in all cases or does it have to be enabled explicitly?

Reply:
Yes, all these steps shall be done prior to copying from a peer memory. However, without enabling peer access, the d2d copy will still work but with terrible performance.

Reply:
Should OMPI enable peer access then? Is there any downside to a library doing that without the application knowing?

Reply:
I think I am confused. The op, does that truly transfer data between two different GPUs? Is this IPC communication? I assumed it is a data transfer between two buffers but on the same GPU

Reply:
We could end up with two buffers on two different devices (probably not in `MPI_Reduce` but through a call to `MPI_Reduce_local`). In the vast majority of cases this does not matter so optimizing is not critical.

Reply:
ok, I am not sure how. The current accelerator framework design only allows for one GPU per MPI process. So in that case, if there is a second GPU and you want to access data from that GPU, that would mean that you have to use the IPC mechanism (which does the peer-enable stuff internally as far as I know, at least for ROCm)

Reply:
That is a limitation that will hopefully change at some point. I'd rather cover all bases here than run into it once we support more than one device.

Anyway, since this is a corner case I will not change this. Patches welcome if you feel this corner case is important. I don't have more time for this.

Reply:
Thinking more about this I realized that this entire logic needs to be changed. I see three cases:
1. data located on different GPU belonging to the same process: manually copying the data upfront is bad for performance, GPUs are really good at doing this automatically, especially in the context of the same process.
2. data located on different GPU belonging to the different processes: we don't cover that case yet as it will require different reduction algorithms (as this capability would remove one explicit communication).
3. data located on main memory: here we only need to explicitly copy if the GPU does not have direct access to the data. We can determine this using the VMM patch that made it into main few days ago.